src/java/org/apache/cassandra/net/MessagingService.java

```diff
@@ -766,24 +766,25 @@ public void shutdown()
         // We may need to schedule hints on the mutation stage, so it's erroneous to shut down the mutation stage first
         assert !StageManager.getStage(Stage.MUTATION).isShutdown();
 
-        // the important part
-        if (!callbacks.shutdownBlocking())
-            logger.warn("Failed to wait for messaging service callbacks shutdown");
-
         // attempt to humor tests that try to stop and restart MS
         try
         {
             clearMessageSinks();
-            for (SocketThread th : socketThreads)
+            for (SocketThread th : socketThreads) {
                 try
                 {
+                    // Close incoming connections
                     th.close();
                 }
                 catch (IOException e)
                 {
                     // see https://issues.apache.org/jira/browse/CASSANDRA-10545
                     handleIOException(e);
                 }
+            }
+            // Wait to finish callbacks before closing outbound connections
+            if (!callbacks.shutdownBlocking())
+                logger.warn("Failed to wait for messaging service callbacks shutdown");
```

Comment:
Nit: can we make this log and comment slightly more clear? The first time I read this, I thought it was waiting for callbacks to complete but it's actually just shutting down the reaper executor service that expires callbacks? Or is there some synchronization I might have missed?

Reply:
I think it is doing what the comment + log say. It's just not obvious due to ExpiringMap's implementation:

in ExpiringMap#put()
```
       if (shutdown)
        {
            // StorageProxy isn't equipped to deal with "I'm nominally alive, but I can't send any messages out."
            // So we'll just sit on this thread until the rest of the server shutdown completes.
            //
            // See comments in CustomTThreadPoolServer.serve, CASSANDRA-3335, and CASSANDRA-3727.
            Uninterruptibles.sleepUninterruptibly(Long.MAX_VALUE, TimeUnit.NANOSECONDS);
```

Reply:
Ah. So it also turns off the "ability to add callbacks", on top of shutting down the reaper. This is ... weirdly implemented